core/domain/question_fetchers.py

```diff
@@ -108,6 +134,31 @@ def get_questions_by_ids(
     return questions
 
 
+def get_multiple_questions_by_ids_and_version(
+    question_ids_and_versions: List[Tuple[str, Optional[int]]]
+) -> List[Optional[question_domain.Question]]:
+    question_models = question_models.QuestionModel.get_version_multi(
+        question_ids_and_versions)
+    return [
+        get_question_from_model(question_model)
+        for question_model in question_models
+        if question_model is not None else None]
+
+
+def get_question_ids_by_skill_ids(skill_ids: List[str]) -> Dict[str, List[str]]:
```

Comment:
Not sure if this will give you a balanced list of questions. AFAICT the get_question_skill_links_by_skill_ids() method sorts everything by last_updated and returns the most recent data. So if we create all the questions for one skill first and then create all the questions for another, you might only get the second one.

You might need another method in the storage layer to handle the query you want in a more balanced way (fetching at most X questions for each skill), if it doesn't already exist. Not really sure how to do that without making one query for each skill ID, though.

Reply:
Ah ok, if you actually want all, then you might probably need a paged approach since I believe we already have > 1000 questions. (And in any case, you might need to shard into bundles of e.g. 20 skills, anyway -- if you look at the impl of get_question_skill_links_by_skill_ids() it uses an ndb "IN" filter and I've never tested it with 100+ items at a time, so I'm not sure what's the max number of skill IDs it will support.)

OK with testing it out as-is if you like but I suspect there is some hidden limit.

Reply:
Per conversation with Ben -- paged approach is OK.